dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc

```diff
@@ -59,88 +72,149 @@ NvJpeg2000DecoderInstance::~NvJpeg2000DecoderInstance() {
-                                               DecodeParams opts,
-                                               const Context &ctx) {
-  void *pixel_data[NVJPEG_MAX_COMPONENT] = {};
-  size_t pitch_in_bytes[NVJPEG_MAX_COMPONENT] = {};
-  int64_t pixel_size = dali::TypeTable::GetTypeInfo(ctx.pixel_type).size();
-  int64_t component_byte_size = ctx.shape[0] * ctx.shape[1] * pixel_size;
+nvjpeg2kImage_t NvJpeg2000DecoderInstance::PrepareOutputArea(uint8_t *out,
+                                                             void **pixel_data,
+                                                             size_t *pitch_in_bytes,
+                                                             int64_t output_offset_x,
+                                                             int64_t output_offset_y,
+                                                             const Context &ctx) {
+  const int64_t pixel_size = dali::TypeTable::GetTypeInfo(ctx.pixel_type).size();
+  const int64_t component_byte_size = ctx.shape[0] * ctx.shape[1] * pixel_size;
+  const int64_t offset_byte_size = (ctx.shape[1] * output_offset_y + output_offset_x) * pixel_size;
   for (uint32_t c = 0; c < ctx.shape[2]; c++) {
-    pixel_data[c] = out + c * component_byte_size;
+    pixel_data[c] = out + c * component_byte_size + offset_byte_size;
     pitch_in_bytes[c] = ctx.shape[1] * pixel_size;
   }
 
-  nvjpeg2kImage_t output_image;
-  output_image.pixel_data = pixel_data;
-  output_image.pitch_in_bytes = pitch_in_bytes;
-  output_image.num_components = ctx.shape[2];
-  output_image.pixel_type = ctx.pixel_type == DALI_UINT8 ? NVJPEG2K_UINT8 : NVJPEG2K_UINT16;
+  nvjpeg2kImage_t image;
+  image.pixel_data = pixel_data;
+  image.pitch_in_bytes = pitch_in_bytes;
+  image.num_components = ctx.shape[2];
+  image.pixel_type = ctx.pixel_type == DALI_UINT8 ? NVJPEG2K_UINT8 : NVJPEG2K_UINT16;
 
-  auto ret = nvjpeg2kDecode(nvjpeg2k_handle_, *ctx.nvjpeg2k_decode_state,
-                            *ctx.nvjpeg2k_stream, &output_image, *ctx.cuda_stream);
+  return image;
+}
 
-  if (ret == NVJPEG2K_STATUS_SUCCESS) {
-    return true;
-  } else if (ret == NVJPEG2K_STATUS_BAD_JPEG || ret == NVJPEG2K_STATUS_JPEG_NOT_SUPPORTED) {
-    return false;
+bool NvJpeg2000DecoderInstance::DecodeJpeg2000(ImageSource *in, uint8_t *out, const Context &ctx) {
+  // allocating buffers
+  void *pixel_data[NVJPEG_MAX_COMPONENT] = {};
+  size_t pitch_in_bytes[NVJPEG_MAX_COMPONENT] = {};
+
+  if (!ctx.roi) {
+    auto output_image = PrepareOutputArea(out, pixel_data, pitch_in_bytes, 0, 0, ctx);
+    auto ret = nvjpeg2kDecode(nvjpeg2k_handle_, ctx.nvjpeg2k_decode_state,
+                              ctx.nvjpeg2k_stream, &output_image, ctx.cuda_stream);
+    return handle_decode_ret_status(ret, in);
   } else {
-    CUDA_CALL_EX(ret, in->SourceInfo());
-    DALI_FAIL("Unreachable");  // silence a warning
+    // Decode tile by tile: nvjpeg2kDecodeImage seems to be bugged
```

Comment:
```suggestion
    // Decode tile by tile: nvjpeg2kDecodeImage currently doesn't work with tiled jpeg2000
```

Reply:
That is not the case, I couldn't get it to work with images including just one tile.